### PR TITLE
Fix ChainStorageError after a reorg with new block

### PR DESCRIPTION
**Pull request by hansieodendaal**
## Description
Fix ChainStorageError after a reorg with new block

- Fixed a ChainStorageError that occurs due to invalid transactions remaining in the mempool when the 1st block is requested for mining via GRPC directly after a chain reorg.
- Added a failing cucumber test for the above condition (fixed now): `Mempool clearing out invalid transactions after a reorg`.
- Minor cucumber improvements.
- ~~Fixed flaky test `tx_broadcast_protocol_submit_success`~~

## Motivation and Context
See above

## How Has This Been Tested?
Cucumber test `Mempool clearing out invalid transactions after a reorg`

## Checklist:

* [X] I'm merging against the `tari-script` branch.
* [X] I have squashed my commits into a single commit.

### CODE REVIEW

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs [hunk 1]
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
         false
     }
 
+    /// Remove all current mempool transactions from the UnconfirmedPoolStorage, returning that which have been removed
+    pub fn clear_out_all_mempool_transactions(&mut self) -> Vec<Arc<Transaction>> {
```
Comment (philipr-za) [inline base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: I think the more canonically Rust term for this process would be `drain`

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs [hunk 2]
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
...
+        let mut mempool_txs: Vec<Arc<Transaction>> = Vec::new();
```
Comment (philipr-za) [inline base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: Performance nit: when you know how long the vector is going to be you should ideally use `Vec::with_capacity(N)`

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs [hunk 3]
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
...
+        let mut removed_tx_keys = Vec::new();
+        for (tx_key, ptx) in self.txs_by_signature.iter() {
+            trace!(
+                target: LOG_TARGET,
+                "Found tx with key {} in unconfirmed pool",
+                tx_key.get_signature().to_hex()
+            );
+            mempool_txs.push(ptx.transaction.clone());
+            self.txs_by_priority.remove(&ptx.priority);
```
Comment (philipr-za) [inline base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: I might be misunderstanding something about these collections but it feels unnecessary to do this one entry at a time. You could just drain the hashmap values into the return vector and clear these two collections. Why do this one at a time?

File: base_layer/wallet/tests/transaction_service/transaction_protocols.rs [hunk 4]
```diff
@@ -255,7 +255,7 @@ async fn tx_broadcast_protocol_submit_success() {
     );
     let join_handle = task::spawn(protocol.execute());
 
-    // Fails because there is no transaqction in the database to be broadcast
+    // Fails because there is no transaction in the database to be broadcast
```
Comment (philipr-za) [inline base_layer/wallet/tests/transaction_service/transaction_protocols.rs]: I believe that SW's PR here https://github.com/tari-project/tari/pull/2899 also addresses this issue

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs (hunk 3 above)
Comment (hansieodendaal) [reply base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: Tried to do it as a batch, but had some runtime issues. Will look at it again.

File: base_layer/wallet/tests/transaction_service/transaction_protocols.rs (hunk 4 above)
Comment (hansieodendaal) [reply base_layer/wallet/tests/transaction_service/transaction_protocols.rs]: Noted.

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs [hunk 5]
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
...
+        for tx_key in &removed_tx_keys {
+            self.txs_by_signature.remove(&tx_key);
+        }
+
+        mempool_txs
+    }
+
     /// Remove all published transactions from the UnconfirmedPool and discard all double spend transactions.
-    /// Returns a list of all transactions that were removed the unconfirmed pool as a result of appearing in the block.
+    /// Returns a list of all transactions that were removed from the unconfirmed pool as a result of appearing in the
```
Comment (stringhandler) [inline base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: nit: This comment is out of date. This method does not return anything

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs (hunk 3 above)
Comment (stringhandler) [reply base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: Agree with Phil here. This is a really complicated way of clearing out two variables. 

```
let mempool_txs = self.txs_by_signature.drain(..).map(....).collect();
self.txs_by_priority = Vec::new();
return mempool_txs;
```

**Review by stringhandler — APPROVED** [review]
Some changes needed to be followed up in further PRs, but this can go in as is for now

File: base_layer/wallet/tests/transaction_service/transaction_protocols.rs (hunk 4 above)
Comment (hansieodendaal) [reply base_layer/wallet/tests/transaction_service/transaction_protocols.rs]: Fix removed

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs (hunk 2 above)
Comment (hansieodendaal) [reply base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: Code changed as suggested below, not applicable anymore.

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs (hunk 3 above)
Comment (hansieodendaal) [reply base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs]: Done

**Comment by hansieodendaal** [comment]
> Some changes needed to be followed up in further PRs, but this can go in as is for now

All changes effected.

**Comment by hansieodendaal** [comment]
See #2969 where the failing `Scenario: Pruned mode reorg past horizon` test has been made predictable.

**Pull request merged**